Makefile

```diff
@@ -1,27 +1,57 @@
-DOCKER_TAG = dev
-
-deps:
-	pip install --upgrade pip
-	pip install --upgrade wheel
-	pip install --use-wheel -r requirements-dev.txt
-	npm install
-
-dev:
+# Run the development web server.
+.PHONY: dev
+dev: .pydeps .jsdeps
 	PYRAMID_RELOAD_TEMPLATES=1 gunicorn --reload "bouncer.app:app()"
 
+# Run all the tests (Python and JavaScript).
 .PHONY: test
-test:
-	@pip install -q tox
-	tox
+test: .jsdeps pytests
 	./node_modules/.bin/eslint bouncer/scripts
 	./node_modules/karma/bin/karma start karma.config.js
 
+# Run the tests and then print out a coverage report.
+# `make coverage` always re-runs the tests.
+.PHONY: coverage
+coverage: test tox
+	tox -e coverage
+
+# Send a coverage report Codecov based on the existing .coverage file.
+# The Python tests will be run to generate the .coverage file if it doesn't
+# exist, but if the .coverage file is out of date (code has changed) the tests
+# won't be re-run.
+.PHONY: codecov
+codecov: .coverage tox
+	tox -e codecov
+
+# Build the docker image.
 .PHONY: docker
 docker:
-	git archive HEAD | docker build -t hypothesis/bouncer:$(DOCKER_TAG) -
+	git archive HEAD | docker build -t hypothesis/bouncer:dev -
 
 .PHONY: clean
 clean:
 	find . -type f -name "*.py[co]" -delete
 	find . -type d -name "__pycache__" -delete
-	rm -f node_modules/.uptodate
+	-rm -f .pydeps .jsdeps .coverage
```

Comment:
I've replaced `node_modules/.uptodate` with `.jsdeps` to make it consistent with `.pydeps`. This does mean that if the user deletes the entire `node_modules` dir then `make` won't run `npm install` though, so maybe I should change it back. (`.pydeps` has the same issue but for Python it's not an easy fix as we can't as easily know where the user's venv is)